### PR TITLE
Security changes

### LINKED ISSUE #3: Solutions branch
It would be nice if the workshops FAC produced could be completed by people who weren’t here. A solutions branch would be useful for them to check their solutions against an exemplar.

### LINKED ISSUE #5: Providing test cases?
Did you find that having students specify their own test cases worked well? Did they cover all (or a sufficient number of) edge and failure cases?

If not, the alternative could be to provide a complete test suite for them. Their job would then only be to write enough code to result in a passing suite.

### LINKED ISSUE #8: Callback trap in example
If you provide a callback interface to a function, you should ensure it behaves as the user would expect. Namely, the callback should be executed asynchronously, after the current call stack is emptied.

You current example looks like this:
```js
const foo = (val, cb) => {
  return val === 0 
    ? cb(new Error('called with zero')
    : cb(null, 'success');
};
```
This means if my calling code looks like this (stupid example):
```js
const meta = { prefix: null };

foo(1, (err, msg) => console.log(err ? err : `${meta.prefix}: ${msg}`));

meta.prefix = 'Message';
```
I would expect the following output:
```
Message: success
```
However, I actually get:
```
null: success
```

This is because the callback is executed synchronously, before I get around to setting the `meta.prefix` option.

To fix this, use [`process.nextTick`](https://nodejs.org/api/process.html#process_process_nexttick_callback_args):
```js
const foo = (val, cb) => {
  return val === 0 
    ? process.nextTick(cb, new Error('called with zero'))
    : process.nextTick(cb, null, 'success');
};
```

### LINKED ISSUE #10: Contextualise the validation exercises with security issues
The whole exercise is about validating input. 

Given that in foundersandcoders/master-reference#263 we agreed to redistribute some security topics throughout the rest of the curriculum, I think it would be a good idea to contextualise the exercise by explaining how validation fits into the security picture.

### PR DESCRIPTION
**Pull request by eliascodes**
**MERGE #12 FIRST**

Fixes #10 
Fixes #8 
Fixes #5 
Fixes #3 

#### Summary
* Make TDD-ing the validation functions the first step of the exercise, but place it in the context of writing a handler that takes user input from a payload.
* Make writing the handler the second step of the exercise (rather than a stretch goal). The hope is that having to write the handler will force them to see how their choice of error handling approach changes their control flow, in a way that isn't accomplished by writing a series of functions that are disembodied from a wider application.
* Remove e-mail validation; I think the main purpose of the validation angle is to give them an awareness and appreciation of why it's important. Other than that, they should mostly just be focusing on how their error-handling affects their control flow. The [intricacies](http://www.regular-expressions.info/email.html) of validating e-mails are (perhaps) a distraction from this.
* Add acceptance tests. They should be TDD-ing during the exercise, but this set of tests will just make sure they've covered a reasonable number of the edge-cases.
* Add solution.

### Comments on the appropriateness of these changes are welcome

@joeylouise @majakudlicka these are changes for the week you're leading, so your reivews are important.

@njsfield you've indicated you're not maintaining the repo, fair enough, but I've at-ed you since you are the original author and may have thoughts.

@des-des Probably busy with week 7, but if you're interested, have a look.

### CODE REVIEW

**Comment by des-des** [comment]
@eliascodes interested, assign me when ready

File: README.md [hunk 1]
```diff
@@ -45,157 +53,294 @@ console.log(result);
...
-        let body = '';
+## Other Requirements
+* The server should not crash because of a client request.
+* If the client supplies bad data, the server should respond with a `400` (Bad request) status code and a helpful message indicating the problem.
 
-        request.on('data', (data) => {
-            body += data;
-        });
+## Steps
+1. Decide how your validation code will indicate that there is a problem with the payload. You are free to use any of the three approaches outlined above (or use another approach if you feel it is appropriate).
```
Comment (des-des) [inline README.md]: What is wrong with this approach
```js
(req, res) => {
  if (!isValidPayload(req.payload)) return res(400)
  // rest of logic here
}
```

Comment (eliascodes) [inline README.md]: Nothing in principle, although that solution doesn't satisfy the following requirement above:
> If the client supplies bad data, the server should respond with a `400` (Bad request) status code and a helpful message indicating the problem.

Comment (des-des) [inline README.md]: What I mean, is return a bool from the validate func, rather than an error / throw

Comment (eliascodes) [inline README.md]: I understand, what I'm saying is returning a boolean gives you no information about what is wrong with the payload, which is one of the requirements I've set. 

Maybe it's not clear from the instructions that that's what I wanted?

Comment (des-des) [inline README.md]: Nah I think its good

File: README.md [hunk 2]
```diff
@@ -45,157 +53,294 @@ console.log(result);
...
+### Notes
+These notes are important to be aware of in general, but are not necessary for the purposes of the workshop.
 
+#### Error-First Callback Pattern
+There are a couple of gotchas when using this callback pattern:
+* You _must_ ensure that the callback is not called more than once in your function. This can be done either using `if/else` blocks, `switch` statements (with `break`), or early `return` statements (e.g. `return callback(null, result)`).
+* When presenting a callback interface to the caller, it will expect the callback to be executed asynchronously. On Node.js, you can use [`process.nextTick`](https://nodejs.org/api/process.html#process_process_nexttick_callback_args) for this purpose, on the client-side, you can use `setTimeout(Function, 0)`.
```
Comment (joeylouise) [inline README.md]: I think this bit will be confusing for some, but as it's only some extra info then it seems fine to leave it in

Comment (eliascodes) [inline README.md]: OK 👍 

There's no need for you guys to go through this section in workshop anyway, you can just leave it for anyone interested to glance at.
